### ISSUE TITLE
Proposal: Add http_sd_config-compatible endpoint for running integrations

### ISSUE THREAD

**Issue by rfratto** [grafana/agent#1133]
Grafana Agent integrations work by hosting Prometheus metrics at an integration-specific endpoint. For example, with node_exporter enabled, `/integrations/node_exporter/metrics` will expose integrations for those metrics. The returned page is unchanged from the embedded integration, with the original labels and metric names intact. 

Integrations also support a self-scraping mode, which is enabled by default. When self-scraping is enabled for the node_exporter integration, Grafana Agent will configure itself with the following instance config (simplified for brevity):

```yaml
name: integrations/node_exporter
scrape_configs:
- job_name: integrations/node_exporter
  metrics_path: /integrations/node_exporter/metrics
  static_configs:
  - targets: ['<AGENT HOSTNAME:PORT>']
    labels:
      agent_hostname: '<AGENT HOSTNAME:PORT>'
  relabel_configs:
  - action: replace
    target_label: 'instance'
    replacement: '<integration SUO address>'
```

Special integration-specific labels are injected through the combination of static_configs and relabel_configs. While this works well for self-scraping integrations, it prevents non-self scraping users from taking advantage of the same automatic labelling, causing some loss of information, such as the SUO address for an integration. 

I propose that Grafana Agent adds a new API endpoint of `/agent/api/v1/metrics/integrations/sd`. This would return an [http_sd_config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config)-compatible payload with the integration-specific labels present:

```json
[
  {
    "targets": [ "<AGENT HOSTHAME:PORT>" ],
    "labels": {
        "__metrics_path__": "/integrations/node_exporter/metrics",
        "agent_hostname": "<AGENT HOSTNAME:PORT>",
        "instance": "<integration SUO address>"
    }
  }
]
```

This endpoint should support the following query parameters:

* `?integrations`: comma-delimited list of integrations to return. When empty, all running integrations will be returned. The default is the empty string.
* `?scraping`: filters integrations down to integrations for a specific mode. One of `auto` (only return automatically-scraping integrations), `manual` (only non-self scraping integrations), or `all` (return all integrations regardless of self-scraping). Default is `manual`. 

Pros: 

* Labeling consistency between automatic and manual scraping integrations
* Plays well with #1059, where integrations may have non-obvious metric paths by default 
* Simplifies much of the integration manager code 
* Enables Grafana Agent Operator (#883) to discover what integrations are being run for scraping.

Cons:

* Introduces latency to when integrations get discovered. `http_sd_configs` is poll-based and defaults to 60s, but can be lowered.

Out of scope: 

* Changing integrations to take full advantage of this new API endpoint. This would require more dramatic changes to integrations to not require one scrape_config per running integration. See comments in issue for details.

**Comment by rfratto** [comment]
> I then propose that Grafana Agent use this internally for self-scraping integrations.

I don't know if this actually is possible. The config for integrations support both `relabel_configs` and `metric_relabel_configs` which is propagated to the automatically generated scrape job and _not_ the underlying integration. In its current form, it would still require one unique scrape job per running integration. 

It's possible to have the relabel_configs and metric_relabel_configs apply to the integration directly, but it's not clear to me if that would be worthwhile.

**Comment by rfratto** [comment]
> It's possible to have the relabel_configs and metric_relabel_configs apply to the integration directly, but it's not clear to me if that would be worthwhile.

After a little bit of investigation, it would be possible to apply relabel_configs and metric_relabel_configs directly to the integration:

* Relabel configs could be applied once when generating the target to expose via this new API endpoint 
* Metric relabel configs could be performed through a custom implementation of `prometheus.Gatherer`, which filters metrics at scrape time. 

Again, not sure if this is worth it just for the agent to internally use its own endpoint, but doing this would bring even more parity to self-scraping and non-self-scraping.

**Comment by rfratto** [comment]
I'm going to say that using this new API internally is out of scope, but might be a change later on. The change we _would_ need to make is to stop using the default relabel_configs and move those values to the target which would also be exposed over the new API.

**Comment by rfratto** [comment]
Closed by #1142
